### PR TITLE
Implement Send for most of the types

### LINKED ISSUE #15: It's safe to implement `Send` for most of the types
I know this has been discussed before, but back then it was conflated with other topics. This issue is focused on [`Send`].

@kenba, you've made sure that `opencl3` is a safe wrapper around an OpenCL FFI. For example the types are not `Copy` or `Clone` to have these safety guarantees as good as possible. Following the [Rust API Guidelines about interoperability], those types should then implement `Send`.

Let me take [`CommandQueue`] as an example. It is a struct with two fields, one is of type [`cl_command_queue`], the other one is of type `cl_uint`. `cl_uint` is not of concern, as it is just a type alias for `i32`, which is already safe. The problem is `cl_command_queue`, which is a raw `*mut c_void` pointer. `CommandQueue` owns the pointer, so no one else outside of `CommandQueue` can manipulate it, once passed in. When using the library the pointer will stay around, you already built the library so that is the case. Given all these things, the `CommandQueue` qualifies as `Send`, hence can implement `Send`, despite having a raw (non-`Send`) pointer internally.

Another good read in regards to FFIs and `Send` is [this answer on Stack Overflow].

[`Send`]: https://doc.rust-lang.org/std/marker/trait.Send.html
[`CommandQueue`]: https://github.com/kenba/opencl3/blob/d4c2055438778600ffab344e4d59f94d519a58a7/src/command_queue.rs#L39-L42
[`cl_command_queue`]: https://docs.rs/opencl3/0.2.1/opencl3/types/type.cl_command_queue.html
[Rust API Guidelines about interoperability]: https://rust-lang.github.io/api-guidelines/interoperability.html#types-are-send-and-sync-where-possible-c-send-sync
[this answer on Stack Overflow]: https://stackoverflow.com/questions/50258359/can-a-struct-containing-a-raw-pointer-implement-send-and-be-ffi-safe/50341075#50341075

### PR DESCRIPTION
**Pull request by vmx**
This way it's possible to move the execution off a main thread.

Closes #15.

### CODE REVIEW

**Comment by vmx** [comment]
I targeted the `main` branch. If it should go into the `develop` branch, let me know and I'll redo this PR.

File: src/event.rs [hunk 1]
```diff
@@ -32,6 +32,8 @@ impl Drop for Event {
     }
 }
 
+unsafe impl Send for Event {}
```
Comment (kenba) [inline src/event.rs:35]: `ExecuteKernel` is designed to be used as "builder" for the `clEnqueueNDRangeKernel` command.  
I cannot see a reason to pass an `ExecuteKernel` between threads, can you?

File: src/memory.rs [hunk 2]
```diff
@@ -717,6 +723,8 @@ impl Drop for Pipe {
     }
 }
 
+unsafe impl Send for Pipe {}
```
Comment (kenba) [inline src/memory.rs:726]: `Pipe`s are just created to pass data between running `Kernel`s on a device, it should not have `Send`.

File: src/program.rs [hunk 3]
```diff
@@ -90,6 +90,8 @@ impl Drop for Program {
     }
 }
 
+unsafe impl Send for Program {}
```
Comment (kenba) [inline src/program.rs:93]: `Program`s are intermediate objects that are used to create `Kernel`s.  
I cannot see  a need to send them between threads, can you?

File: src/svm.rs [hunk 4]
```diff
@@ -35,6 +35,8 @@ struct SvmRawVec<'a, T> {
     fine_grain_buffer: bool,
 }
 
+unsafe impl<'a, T: Send> Send for SvmRawVec<'a, T> {}
```
Comment (kenba) [inline src/svm.rs:38]: This is component of `SvmVec`,  but `SvmVec` is not `Send`, is this correct?

File: src/svm.rs [hunk 5]
```diff
@@ -315,6 +317,8 @@ struct RawValIter<T> {
     end: *const T,
 }
 
+unsafe impl<T: Send> Send for RawValIter<T> {}
```
Comment (kenba) [inline src/svm.rs:320]: As above, this is also a part of the `SvmVec` implementation,  but `SvmVec` is not `Send`, is this correct?

**Review by kenba — CHANGES_REQUESTED** [review]
Thank you Volker.

Unfortunately, we disagree on the scope of the changes:  
you want to provide `Send` for everything;  
while I'm only willing to provide `Send` for types where there is a good reason for it.  

I've commented on the `Send`s that I do not agree with.  
If you have a good reason for implementing them, then please state your reason in the comments, otherwise please remove them.

Also, as you spotted, please can you merge into `develop` instead of `main`.

File: src/svm.rs (hunk 4 above)
Comment (vmx) [reply src/svm.rs:38]: `SvmVec` is automatically `Send`. If a type consists only of types that are `Send`, then that type is automatically `Send`. This is the reason why the [Interoperability API Guidelines](https://rust-lang.github.io/api-guidelines/interoperability.html#types-are-send-and-sync-where-possible-c-send-sync) say that types should be `Send` where possible. And "where possible" is about "thread safety characteristics", and I think all the types here are safe in that regard.

File: src/svm.rs (hunk 5 above)
Comment (vmx) [reply src/svm.rs:320]: See https://github.com/kenba/opencl3/pull/16#discussion_r641981784.

File: src/program.rs (hunk 3 above)
Comment (vmx) [reply src/program.rs:93]: I thought I need it (my current code does), but I revisited it and it turns out I don't actually need it. Thanks for making me having another look.

**Comment by vmx** [comment]
I've opened https://github.com/kenba/opencl3/pull/17 which removes `Send` from `ExecuteKernel`, `Pipe` and `Program`.

I disagree and still believe that `Send` should be implemented where technically possible as it's about safety guarantees and not about how people should use a certain API. Though I respect your decision @kenba and I'm happy that we were able to find agreement on the types that I need for my code, so that I don't have to implement `Send` manually.

**Pull request closed without merging**